Add gscan component

kinow opened the pull request:
These changes partially address #139

Initial version of the `GScan` component. Will need more iterations later until the requirements in #139 are fully addressed.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

kinow commented:
Note: it takes users to `/workflows/$workflow-name`. Which at the moment is rendered by the `Tree` view. Later, we should replace it by a `Workflow.vue` I think, which adds the initial area to add tabs as in #98

kinow commented:
And a GIF of what it looks like in action:

![GIFrecord_2019-11-05_165021](https://user-images.githubusercontent.com/304786/68177225-8d547d00-ffec-11e9-9b73-a67781bc61ce.gif)

kinow commented:
Build failed due to sass-loader and vuetify. Removing node-sass to see if it will pass now (off to home, will log in again later to check if build passed). Only tested with the workflow five, so it would be good if a reviewer could test with other workflows and see if there's anything broken :)

kinow commented:
Build passed on Travis :tada:

oliver-sanders commented on src/App.vue:
```diff
@@ -7,13 +7,106 @@
+
 const defaultLayout = 'empty'
 
+const QUERIES = {
```
Why is this stuff here rather than in `GScan.vue`?

kinow replied:
I got used to having a component rendering, and the data fetching/massaging elsewhere. As it seems to make the components easier to maintain.

So for `GScan`, I tried to create it with only props, and then plug the data in. That's when I realized I didn't have a view/route for GScan, as it is displayed in pretty much every view of the system. It is also easier to write unit tests and check the coverage.

But if you find it confusing or that it is not how it was supposed to work with the WorkflowService, we can move it to `GScan.vue`. As that component it used throughout the application maybe it makes sense that it both fetches data and also renders it.

kinow replied:
Moving it to `GScan.vue`. If necessary we can move it to somewhere else later 👍

oliver-sanders commented on src/App.vue:
```diff
@@ -7,13 +7,106 @@
+            meanElapsedTime
+            name
+          }
+          jobs(sort: { keys: ["submit_num"], reverse:true }) {
```
Should be able to the summary thinggy (is it `stateSummary`) rather than requesting full info on each job.

kinow replied:
Just had a look at the schema in `cylc.flow.network.schema`, the closest I found was `stateTotals`.

![image](https://user-images.githubusercontent.com/304786/68340446-5ba5f800-014b-11ea-8527-8c3cd6bd87e1.png)

We can use that to give the numbers - which are missing - to the tooltip. I think there was a recent discussion on riot where @hjoliver and @dwsutherland mentioned something changes to the backend that I think is the same thing you are suggesting here. (I think.)

oliver-sanders replied:
Here's the endpoint we used to call:

https://github.com/cylc/cylc-flow/blob/e3deb7d7049117e7c2a487f558f8b839a7d2ab25/lib/cylc/network/httpserver.py#L328

From memory I think it gave state totals and the five most recent task-jobs in each state?

kinow replied:
That'd be handy for the UI. Maybe we should open a new issue in Cylc UI Server - I think - to work on this?

hjoliver replied:
This has been resolved below, with some follow-up needed. I'll update the main gscan issue #139 description.

oliver-sanders commented on src/components/cylc/GScan.vue:
```diff
@@ -0,0 +1,102 @@
+      <div v-for="workflow in workflows" :key="workflow.id">
+        <v-list-item :to="`/workflows/${ workflow.name }`">
+          <v-list-item-action>
+            <!-- TODO: the workflows are always running now, but later we will need https://github.com/cylc/cylc-ui/issues/90 adn update the icon displayed here -->
```
Note workflows can be running but held.

kinow replied:
Aaahhh, good point! I added this icon to fill in the space as otherwise the text would be left-aligned without anything before it.

Could we leave this icon for now? I will work on #90 after this, and update GScan to use the new component, which should handle running, held, and.... stopped (?).

oliver-sanders replied:
Should be really easy to implement, the field `workflow.status` field is either `running` or `held`:

```graphql
query {
  workflows{
    status
  }
}
```

hjoliver replied:
@kinow - I think you've addressed this, right? (gscan held icon).

oliver-sanders commented on src/components/cylc/GScan.vue:
```diff
@@ -0,0 +1,102 @@
+            <v-layout align-center align-content-center wrap>
+              <v-flex grow>{{ workflow.name }}</v-flex>
+              <v-flex shrink ml-4>
+                <!-- task summary tooltips -->
```
Nice.

I think we might want to factor out some of this functionality into a common component so all views can benefit from these features (e.g. tooltips).

I was thinking of something like `JobBox` which either holds:
* State totals (as in this case).
* A collection of jobs (which collapses down to state totals if there are too many jobs).

Food for later thought...

hjoliver replied:
@oliver-sanders - can you post an issue for this, if you think it's needed.

oliver-sanders commented on src/App.vue:
```diff
@@ -7,13 +7,106 @@
+const QUERIES = {
+  root: `
+    {
+      workflows {
```
BTW: I think we will need to improve the whole `gproxy` thing so it can merge this query properly, my bad I should have opened a follow on issue at the time.

kinow replied:
There are two or three issues created that involve changes in the `gproxy`, so there's room for improvement, but it was a life saver. Without that early in the project, we would have a lot built on top of - probably - multiple queries spread over the project.

hjoliver replied:
@oliver-sanders - ditto this - is an issue needed, it will it be obvious later on?

kinow commented:
Rebased, fixed conflicts. May still have to update the code based on further comments from @oliver-sanders (I think I replied with only more questions hehe)

hjoliver commented:
@kinow - have you tested this with more than one workflow?  When I do, the first one I click on (in the scan bar) appears in the tree view, but then second one results in a blank view.  The first view comes back if I click on the first one again in the sidebar.

kinow commented:
> @kinow - have you tested this with more than one workflow? When I do, the first one I click on (in the scan bar) appears in the tree view, but then second one results in a blank view. The first view comes back if I click on the first one again in the sidebar.

Oh! Interesting issue! I surely tested two workflows, `five`, and `families2` (an example you provided in some issue). But I probably did not test both workflows at the same time :smile: 

Let me try it and see what happens.

kinow commented:
I think it's fixed now!

- moved the logic of computing the workflow out of Vuex, into the view Tree.vue
- fixed the router title problem, which was not displaying the workflow name in the toolbar
- tested with two workflows at the same time
    - one workflow (`families2` source below) occasionally had some exceptions due to a value missin from the `lookup` table (i.e. a task `firstParent` was missing from `familyProxies`). After that, the application wouldn't allow users to switch workflows, being piratically broken. Added a try/catch and will log the exception to the browser console, and allow users to keep using the rest of the app - there may be a moment that the workflow tree is empty due to this exception

kinow commented:
(just need to fix the tests)

hjoliver commented:
That's got it, looks great.

kinow commented:
>Finally, we need some way to access to the cytoscape graph view!

Without #98 there's no way to access the graph now :disappointed: unless one knows the URL. Do you have any suggestion where we could include a link for the Graph view? Or perhaps we should try to display both components at the same time by default?

hjoliver commented:
As a temporary measure, could we just keep the old scan table view via a link from the dashboard?

kinow commented:
> As a temporary measure, could we just keep the old scan table view via a link from the dashboard?

Sure. Or should we add that link under the Dashboard entry???? It just occurred to me that I could have simply added it under the existing values (a bit of d'oh moment). If that's fine, it'll be a 5 min job.

hjoliver commented:
Yep that’s fine 👍🏻

kinow commented:
> Yep that’s fine 👍🏻

Done

![image](https://user-images.githubusercontent.com/304786/68542038-6c38c580-040c-11ea-8c53-314fe5a36640.png)

kinow commented on src/components/cylc/Drawer.vue:
```diff
@@ -36,19 +36,22 @@
         </v-list-item-action>
         <v-list-item-title
-          v-text="link.text"
+            v-text="link.text"
```
Hm, I've added spaces to a few lines here without actually changing anything. Let me fix these.

kinow commented on src/components/cylc/GScan.vue:
```diff
@@ -0,0 +1,204 @@
+import { workflowService } from 'workflow-service'
+
+const QUERIES = {
+  root: `
```
@oliver-sanders 👆

kinow commented on src/components/cylc/GScan.vue:
```diff
@@ -0,0 +1,204 @@
+          return 'mdi-pause-octagon'
+        default:
+          return 'mdi-help-circle'
+      }
```
@oliver-sanders, @hjoliver also added an extra commit to choose the workflow icon. Given we have the icons loaded anyway, and they fit the CSS styles, would it be fine? Or do we need the component with SVG? They don't look exactly like the design sketch, but looks good to me.

![image](https://user-images.githubusercontent.com/304786/68551444-e8192900-0471-11ea-96ec-2bcc85205e54.png)

(did a `cylc hold five` to test the other icon)

kinow replied:
(note that the Workflows menu item is activated as it matches the "/workflows*" route)

hjoliver replied:
They play icons look fine to me, and we can change to custom svg later on if we want to.

hjoliver commented:
I'm seeing a faint partial border (just the corners) around the task status icons though:
![shot](https://user-images.githubusercontent.com/2362137/68553383-7cd95200-0485-11ea-9511-b2f234296c2f.png)

kinow commented:
> 
> 
> I'm seeing a faint partial border (just the corners) around the task status icons though:
> ![shot](https://user-images.githubusercontent.com/2362137/68553383-7cd95200-0485-11ea-9511-b2f234296c2f.png)

Your eyes are 20/20. I might need to visit my optician haha. Let me try to fix this, hopefully something simple. Thanks!

hjoliver commented:
> Your eyes are 20/20. 
 
Maybe my glass are :eyeglasses:

kinow commented:
Interesting, looks like there's nothing on Chrome, but on Firefox I can see these border, just like in your screenshot 😕

kinow commented:
Fixed (hopefully!). Gotta really buy some 🍻 to whoever worked on those browser developer tools. Found which class exactly was causing the element to render the borders. It was the `elevation-0`. I thought it would remove the elevation (buttons have have a fading grey border underneath it, to give the effect of elevation).

kinow commented:
One more commit, that is just adding some `if (Object.hasOwnProperty.call(object, 'property))` for the graph data.

We've added back the graph link, but the graph was broken due to properties being accessed before the object was populated/synced in/with Vuex store. Now it should

- add the gscan component under the existing menu items in the drawer/left sidebar
- fix the workflow title in the toolbar when browsing the tree/graph views
- fix the graph view

hjoliver commented:
Looking good.

Now wondering if gscan tool-tip reduction is feasible this week (or did that require new data from the back-end? I forget the result of that discussion...).

New:
![shot](https://user-images.githubusercontent.com/2362137/68561386-364a1e80-04aa-11ea-8df7-3029abbdc780.png)

Old:
![shot2](https://user-images.githubusercontent.com/2362137/68561638-46163280-04ab-11ea-8e9b-a7de8ed6147b.png)

kinow commented:
>Now wondering if gscan tool-tip reduction is feasible this week (or did that require new data from the back-end? I forget the result of that discussion...).

I think @oliver-sanders pointed the [http endpoint](https://github.com/cylc/cylc-flow/blob/e3deb7d7049117e7c2a487f558f8b839a7d2ab25/lib/cylc/network/httpserver.py#L328) from where we used to get this data from, and in the discussion here and on Riot we were discussing that it would be possible to get state totals from the backend - though I believe it is not the same problem.

I think [this is the code in GScan that creates the message](https://github.com/cylc/cylc-flow/blob/de7d938496e82dbdfb165938145670dd8e801efd/lib/cylc/gui/gscan.py#L852) from your screenshot. It gets the data from [this method from the `state_summary_mgr`](https://github.com/cylc/cylc-flow/blob/de7d938496e82dbdfb165938145670dd8e801efd/lib/cylc/state_summary_mgr.py#L204).

![image](https://user-images.githubusercontent.com/304786/68562761-e53d2900-04af-11ea-860b-a5d4f728efe0.png)

I will try to reproduce this in JS :point_up:

hjoliver commented:
Thanks @kinow (and tomorrow is OK!)

kinow commented:
Rebased, and implemented:

- sorting the tasks within each state alphabetically (JS, when creating the summary data)
- display only N (default to 5) tasks (JS/Vue in the component, we use `array.slice(0, N)` to get up to N... then another `<span v-if="tasks.length > N">` that displays the text "And ${tasks.length - N} more" as before.

hjoliver commented:
Nice, thanks @kinow.  Looks great, works great!

We'll need sorting by "time" not name (e.g. when looking at succeeded tasks, want the most recently succeeded ones first).  But that can be a follow-up. This is great for my demo as-is (I won't tell anyone it's not really "most recent", he heh :grin: )

The Cylc 7 version also does this server side (I had forgotten) but it is probably better in principle that the UI does it (some future UI component might need to list all tasks in a state, not just a few.)

hjoliver commented:
@oliver-sanders - I'm going to merge this without your approval, as I need it in for demos soon. I believe all your comments have been addressed above, with some things punted to other issues (and/or updated the gscan Issue description). Note two questions on things that might need a new Issue page.
